### PR TITLE
dialects: (hw) Implement HW dialect starting with InnerRefAttr

### PR DESCRIPTION
**Pull request by lucjaulmes**
I think this is a decent first chunk for `HW`. Wider context available in #1704

### CODE REVIEW

File: tests/filecheck/dialects/hw/hw_ops.mlir [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// RUN: XDSL_ROUNDTRIP
```
Comment (superlopuh) [inline tests/filecheck/dialects/hw/hw_ops.mlir:1]: could you please add an `XDSL_ROUNDTRIP_GENERIC` also?

File: xdsl/dialects/hw.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string("<@")
+        printer.print_string(self.module_ref.root_reference.data)
+        printer.print_string("::@")
+        printer.print_string(self.sym_name.data)
+        printer.print_string(">")
+
+
+HW = Dialect(
+    [],
```
Comment (compor) [inline xdsl/dialects/hw.py:79]:
```suggestion
HW = Dialect(
    "hw",
    [],
```

**Review by compor — APPROVED** [review]

**Comment by compor** [comment]
@lucjaulmes can you please update and have all CI tests pass?
This LGTM, but I'm still looking into CIRCT wrt #1704 and if there's space to avoid reimplementing functionality.

**Review by superlopuh — APPROVED** [review]

**Review by luisacicolini — APPROVED** [review]
very interesting!

**Comment by lucjaulmes** [comment]
I’m not sure I understand how CI works. The tests/builds failed because of changes in 845a3a9e -- that isn’t in this branch and was likely even pushed to main after my commits were pushed here. That makes results different from running linters/tests locally and a bit hard to predict :raised_eyebrow:

**Pull request merged**